lib/kubernetes-deploy/deploy_task.rb

```diff
@@ -244,14 +250,21 @@ def validate_definitions(resources)
 
     def discover_resources
       resources = []
+      crds = cluster_resource_discoverer.crds(@sync_mediator).each_with_object({}) { |crd, memo| memo[crd.kind] = crd }
       @logger.info("Discovering templates:")
 
       Dir.foreach(@template_dir) do |filename|
         next unless filename.end_with?(".yml.erb", ".yml", ".yaml", ".yaml.erb")
 
         split_templates(filename) do |r_def|
-          r = KubernetesResource.build(namespace: @namespace, context: @context, logger: @logger,
-                                       definition: r_def, statsd_tags: @namespace_tags)
+          kind = r_def["kind"]
+          r = if crds[kind] && !KubernetesDeploy.const_defined?(kind)
```

Comment:
What made you decide to move this here instead of ` KubernetesResource.build`?

Reply:
I wasn't happy with having to pass a `crd` object into all calls to `KubernetesResource.build`, so wanted a solution where I could avoid that. For this reason, I moved the logic to check for a generic CR outside of `KubernetesResource.build`.

Initially I was going to make a method like `KubernetesResource.build_cr`, but the logic for it is quite sparse, so the same thing can be tersely accomplished here in `deploy_task.rb`. My underlying motivation is that instances of `CustomResource` aren't as concrete as completely specified resources (e.g. `Pod`, `Deployment`), and we can infer when we have one with the simple conditional `if crds[kind] && !KubernetesDeploy.const_defined?(kind)`. To that end, `CustomResource.new` is semantically similar enough to `KubernetesResource.build` that I was ok placing them together. Another way of looking at it is that both `CustomResource` and `KubernetesResource` infer properties of generic objects and return a fully specified resource, which puts them on equal footing from an abstraction point of view.

Said that, I'm ok moving things around if there's disagreement on my approach. At the end of the day the only thing that's important here is that we create `CustomResource` objects when appropriate.